### PR TITLE
Anonymize some line numbers in UI test output

### PR DESCRIPTION
**Pull request by petrochenkov**
New unstable flag `-Z ui-testing` is introduced. This flag changes diagnostic output of the compiler *in some way* making it more suitable for UI testing (this is intentionally vague).
At the moment this flag anonymizes line numbers at line starts thus solving the largest issue with UI test diffs. If diffs continue to be too noisy, some other tweaks could be applied (e.g. anonymizing lines/columns in `--> $DIR/file.rs:line:column`), but this needs some time and experience (we shouldn't diverge too much from the actual output in general).

If comment `// disable-ui-testing-normalization` is added to an UI test, then `-Z ui-testing` is not passed.

Closes https://github.com/rust-lang/rust/issues/46643

### CODE REVIEW

File: src/test/ui/lint-unconditional-recursion.stderr [hunk 1]
```diff
@@ -1,157 +1,157 @@
...
-    |
-    = help: a `loop` may express intention better if this is on purpose
+  --> $DIR/lint-unconditional-recursion.rs:119:5
+   |
+LL |     fn index(&self, x: usize) -> &Baz { //~ ERROR function cannot return without recurring
+   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ cannot return without recurring
+LL |         &self[x]
+   |          ------- recursive call site
+   |
+   = help: a `loop` may express intention better if this is on purpose
```
Comment (petrochenkov) [inline src/test/ui/lint-unconditional-recursion.stderr]: Example of a test with "long" actual numbers that become "short" anonymized numbers.

**Comment by estebank** [comment]
This doesn't deal with the line number in the `file:line:col` section, but should be fine. LGTM.

File: src/test/ui/type-check/issue-40294.stderr [hunk 2]
```diff
@@ -1,19 +1,19 @@
...
    |
-15 | / fn foo<'a,'b,T>(x: &'a T, y: &'b T) //~ ERROR type annotations required
-16 | |     where &'a T : Foo,
-17 | |           &'b T : Foo
-18 | | {
+LL | / fn foo<'a,'b,T>(x: &'a T, y: &'b T) //~ ERROR type annotations required
+LL | |     where &'a T : Foo,
+LL | |           &'b T : Foo
+LL | | {
 19 | |     x.foo();
```
Comment (oli-obk) [inline src/test/ui/type-check/issue-40294.stderr]: There's a line number left here

Comment (petrochenkov) [reply src/test/ui/type-check/issue-40294.stderr]: Yeah, noticed these too late, will fix today or tomorrow.
I haven't figured out yet where they come from.

**Comment by petrochenkov** [comment]
>This doesn't deal with the line number in the file:line:col section

cc https://github.com/rust-lang/rust/issues/46643#issuecomment-350831862

**Comment by petrochenkov** [comment]
Updated.

**Comment by nikomatsakis** [comment]
Now that we have the `//~ ERROR` lines, sgtm. 

@bors r+

**Comment by petrochenkov** [comment]
@bors p=1
The patch is conflict-prone + other PRs in the queue modifying UI tests will benefit from it.

**Comment by kennytm** [comment]
Legit. New UI tests appeared. 

```
[00:53:39] failures:
[00:53:39]     [ui] ui/feature-gate-if_while_or_patterns.rs
[00:53:39]     [ui] ui/issue-45157.rs
[00:53:39]     [ui] ui/lint/lint-group-nonstandard-style.rs
[00:53:39]     [ui] ui/macros/macro-at-most-once-rep-ambig.rs
[00:53:39]     [ui] ui/span/issue-39018.rs
```

@bors r- p=4540 

— This PR will very likely break again if any other PRs merged before it. Please rebase and fix ASAP.

**Comment by petrochenkov** [comment]
Excellent, now https://github.com/rust-lang/rust/pull/48337 broke the core logic.

**Comment by petrochenkov** [comment]
Appveyor 3h timeout
@bors retry

**Comment by Mark-Simulacrum** [comment]
Looks like now 64-bit AppVeyor is starting to hit the timeout as well :/

**Comment by kennytm** [comment]
check-aux is timing out quite a lot actually. Perhaps doing cargo-test + pretty test together is too long for Windows. I’m thinking of moving one of the check-aux tests to the tools builder, but we need to figure out which ones can balance the time.

**Pull request merged**